### PR TITLE
Misc 2022-05

### PR DESCRIPTION
**Pull request by LukeShu**
## Description

Just like the [April PR](https://github.com/emissary-ingress/emissary/pull/4215) that just went in, this PR is a bunch of mostly unrelated things:
- The first 6 commits are of the theme "kill off `pkg/k8s`".
- The 7th commit is addressing feedback from #4215.
- The 8th and 9th commits are addressing woes with automated repatriation seen over at https://github.com/emissary-ingress/emissary/pull/4217.
- The 10th commit is a bugfix I noticed while working on the prior 2 commits.
- The final commit is fixing that #4194 broke `make deploy` (thanks @LanceEa for pointing this out).

Don't be scared by the massive ` 94 files changed, 34398 insertions(+), 2006 deletions(-)`; excluding `vendor/` it's actually a reduction:
```console
$ git diff --stat origin/release/v2.3 origin/lukeshu/misc-2022-05 :!vendor/
…
 35 files changed, 409 insertions(+), 1687 deletions(-)
```

## Related Issues
There'll probably be a "Misc 2022-05 pt2" coming later. But between `make deploy` and repatriation, I figure this would be helpful to land sooner.

## Testing
I tested `make deploy`. I haven't tested the CI/repatriation things, but if they don't work, I'll just be filing pt2 later today!

## Checklist

 - [x] I made sure to update `CHANGELOG.md`. - no applicable changes
   
   Remember, the CHANGELOG needs to mention:
    + Any new features
    + Any changes to our included version of Envoy
    + Any non-backward-compatible changes
    + Any deprecations
 
 - [x] This is unlikely to impact how Ambassador performs at scale.
 
   Remember, things that might have an impact at scale include:
    + Any significant changes in memory use that might require adjusting the memory limits
    + Any significant changes in CPU use that might require adjusting the CPU limits
    + Anything that might change how many replicas users should use
    + Changes that impact data-plane latency/scalability
 
 - [ ] My change is adequately tested.
 
   Remember when considering testing:
    + Your change needs to be specifically covered by tests.
       + Tests need to cover all the states where your change is relevant: for example, if you add a behavior that can be enabled or disabled, you'll need tests that cover the enabled case and tests that cover the disabled case. It's not sufficient just to test with the behavior enabled.
    + You also need to make sure that the _entire area being changed_ has adequate test coverage.
       + If existing tests don't actually cover the entire area being changed, add tests.
       + This applies even for aspects of the area that you're not changing – check the test coverage, and improve it if needed!
    + We should lean on the bulk of code being covered by unit tests, but...
    + ... an end-to-end test should cover the integration points
 
 - [x] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently. - no tricks

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -81,7 +81,7 @@ deploy: push preflight-cluster
 	$(MAKE) deploy-only
 .PHONY: deploy
 
-deploy-only: preflight-dev-kubeconfig $(tools/kubectl) $(OSS_HOME)/manifests/emissary/emissary-crds.yaml $(boguschart_dir)
+deploy-only: preflight-dev-kubeconfig $(tools/kubectl) build-output/yaml-$(patsubst v%,%,$(VERSION)) $(boguschart_dir)
```
Comment (kflynn) [inline Makefile:84]: Is it time to do 

```
VERSION_NO_V = $(patsubst v%,%,$(VERSION))
```

and just make this easier to read everywhere??

**Review by kflynn — APPROVED** [review]
Lovin' it! though I'm mostly trusting the Go compiler on some of the `k8s` -> `kates` stuff. Looks sane, could easily have missed something subtle, all I'm saying. 🙂 

My one question is a fine thing for a future PR. Let's land this!

File: Makefile (hunk 1 above)
Comment (LukeShu) [reply Makefile:84]: I'm not _opposed_ but I'm fine with `$(patsubst v%,%,$(VERSION))` everywhere.

File: builder/builder.mk [hunk 2]
```diff
@@ -268,10 +268,6 @@ ifneq ($(IS_PRIVATE),)
 else
 	@printf '$(CYN)==> $(GRN)recording $(BLU)%s$(GRN) => $(BLU)%s$(GRN) in S3...$(END)\n' "$$(git rev-parse HEAD)" $(patsubst v%,%,$(VERSION))
 	echo '$(patsubst v%,%,$(VERSION))' | aws s3 cp - 's3://$(AWS_S3_BUCKET)/dev-builds/'"$$(git rev-parse HEAD)"
```
Comment (LanceEa) [inline builder/builder.mk:270]: Based on your commit description, you are removing the push to S3 but isn't the copy to S3 done on this line or am I missing something?

File: pkg/kates/apply.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+	"context"
+	"os"
+
+	"k8s.io/kubectl/pkg/cmd/apply"
+	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+)
+
+// IncoherentApply is like `kubectl apply`, and DOES NOT count as a "write" for the purposes of the
+// Client's read/write coherence.
+func (c *Client) IncoherentApply(ctx context.Context, stdio IOStreams, args ...string) error {
```
Comment (LanceEa) [inline pkg/kates/apply.go:13]: You implement this but I didn't see it being called anywhere? is it implementing an interface function that is called as part of the deprecate k8s machinary?

**Comment by LanceEa** [comment]
A couple of small comments/questions to make sure my understanding is correct.

Also, could you comment on the reasoning we are killing off two of the items and the general goal there (just for my understanding):

- pkg/k8s
- pkg/dtest

**Comment by kflynn** [comment]
@LanceEa: `pkg/k8s` and `pkg/kates` have similar purposes, but different history, and it's annoying to distinguish between both. Killing one in favor of the other just makes it easier to talk about things, and easier to maintain.

`dtest` started life as an Emissary thing (IIRC) and moved over to `datawire/dlib.git` so as to be easier to consume in other products, but the Emissary version still had a couple of differences from the `dlib` version. Now that the Emissary code doesn't need those differences, we can stop maintaining the Emissary version, delete it, and just use the `dlib` version.

**Comment by LanceEa** [comment]
@kflynn - That is what I was think was the purpose but wanted to understand

File: builder/builder.mk (hunk 2 above)
Comment (kflynn) [reply builder/builder.mk:270]: This is pushing an entry to the dev-build map in S3 – it's a way to look up the Docker tag pushed for a given GitHub SHA. This is distinct from pushing the chart, which the deletions below here are doing.

(We're going to get rid of the dev-build map, too, but that's a later PR.)

File: pkg/kates/apply.go (hunk 3 above)
Comment (kflynn) [reply pkg/kates/apply.go:13]: It's called from `kubeapply.go` in 4fae46776d028bb2ce204f729196985e805166cd.

Comment (LanceEa) [reply pkg/kates/apply.go:13]: I see it now. Thanks. I knew I was probably just missing it 😄 .

File: builder/builder.mk (hunk 2 above)
Comment (LanceEa) [reply builder/builder.mk:270]: ok, make sense 👍

**Comment by LanceEa** [comment]
lgtm

**Review by LanceEa — APPROVED** [review]
lgtm

**Comment by kflynn** [comment]
Merging so I can rebase another change on this. 🙂

**Pull request merged**